src/periphery/SwapAndDeposit.sol

```diff
@@ -0,0 +1,107 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.20;
+
+import {DNft} from "../core/DNft.sol";
+import {VaultManagerV5} from "../core/VaultManagerV5.sol";
+import {VaultLicenser} from "../core/VaultLicenser.sol";
+import {IVault} from "../interfaces/IVault.sol";
+import {IExtension} from "../interfaces/IExtension.sol";
+import {ERC20} from "@solmate/src/tokens/ERC20.sol";
+import {SafeTransferLib} from "@solmate/src/utils/SafeTransferLib.sol";
+import {ReentrancyGuard} from "@openzeppelin/contracts/utils/ReentrancyGuard.sol";
+
+contract SwapAndDeposit is IExtension, ReentrancyGuard {
+    using SafeTransferLib for ERC20;
+
+    DNft public immutable dNft;
+    address public immutable augustusSwapper;
+    VaultManagerV5 public immutable vaultManager;
+    VaultLicenser public immutable vaultLicenser;
+
+    error NotDNftOwner();
+    error UnlicensedVault();
+    error SwapFailed();
+    error InsufficientAmountOut();
+
+    event SwappedAndDeposited(
+        uint256 indexed tokenId,
+        address indexed tokenIn,
+        address indexed tokenOut,
+        uint256 amountIn,
+        uint256 amountOut,
+        address vault
+    );
+
+    constructor(address _dNft, address _augustusSwapper, address _vaultManager, address _vaultLicenser) {
+        dNft = DNft(_dNft);
+        augustusSwapper = _augustusSwapper;
+        vaultManager = VaultManagerV5(_vaultManager);
+        vaultLicenser = VaultLicenser(_vaultLicenser);
+    }
+
+    function name() external pure override returns (string memory) {
+        return "Swap and Deposit";
+    }
+
+    function description() external pure override returns (string memory) {
+        return "Extension for swapping to a vault's asset and directly depositing into a Note";
+    }
+
+    function getHookFlags() external pure override returns (uint256) {
+        return 0; // No hooks needed for this extension
+    }
+
+    function _swapToCollateral(
+        address tokenIn,
+        uint256 amountIn,
+        uint256 minAmountOut,
+        bytes calldata swapData,
+        address tokenOut
+    ) internal returns (uint256 amountOut) {
+        // Transfer tokenIn from the user to this contract, handling fee-on-transfer tokens
+        uint256 balanceBefore = ERC20(tokenIn).balanceOf(address(this));
+        ERC20(tokenIn).safeTransferFrom(msg.sender, address(this), amountIn);
+        uint256 balanceAfter = ERC20(tokenIn).balanceOf(address(this));
+        uint256 actualAmountIn = balanceAfter - balanceBefore;
```

Comment:
why is all of this necessary? we should just use amountIn for everything.

Reply:
For tokens with fees

Reply:
This is a lot of inefficiency to support fee tokens, would rather we just not support fee on transfer tokens for this scenario tbh. If someone wants to swap their random shitcoins they can do it in two steps.